runtime/compiler/optimizer/StaticFinalFieldFolding.hpp

```diff
@@ -30,6 +30,63 @@
 
 namespace TR { class NodeChecklist; }
 
+/** \brief
+ *
+ * This opt walks the trees and folds direct loads of static final field.
+ *
+ * \details
+ *
+ * This opt walks the trees and tries to fold direct loads of static final field.
+ * For primitive type, it will be folded into a primitive constant. For reference
+ * type, a known object index is assigned to the referenced object, the symref on
+ * the load node is improved to one with the known object index.
+ *
+ * There are two types of static final fields this opt trying to fold:
+ *  1. Fields in special JCL classes that are closely tied to VM core implementation,
+ *     these fields are not expected to be modified once initialized.
+ *     `J9::TransformUtil::foldFinalFieldsIn` contains a list of such classes.
+ *
+ *  2. Other fields that may be modified through reflection or Unsafe APIs.
+ *
+ * For 1, they're always directly folded into a constant. The folding doesn't require any
+ * analysis or protection.
+ *
+ * For 2, they can only be folded with the assumption that they're not to be
+ * modified. However, when modification happens, there JIT code must not execute the code
```

Comment:
I think:  "...happens, the JIT code with the invalid folded value must not be executed."